proposed/SerializerGoals5.0.md

```diff
@@ -0,0 +1,301 @@
+3) Generate code for POCOs that avoids using the unsupported features. This also has higher runtime performance than the loosely-typed fallbacks.
+4) Add a high-performance replacement for targeted scenarios including property getters, property setters, constructors and as a stretch `Type.MakeGenericType and .MakeGenericMethod`.
+5) If possible, add or extend runtime interpreter support for missing cases. Xamarin (and .NET Native) has an interpreter mode.
+6) Throw PlatformNotSupportedException for when there is no alternative.
+
+## Faster startup performance
+
+This is important for scenarios including [micro-services](https://github.com/dotnet/runtime/issues/1568) where processes start and exit frequently. It is also important for benchmarks including TechEmpower.
+
+The serializer has fast startup compared to most other serializers including Newtonsoft's Json.NET, Jil and Utf8Json. Startup perf is mostly affected by generating IL - the Jil serializer is the most aggressive here and has the highest startup cost (but in many cases has the fastest steady-state performance).
+
+There is no stated goal for 5.0 to make the steady-state CPU performance faster than it is now in 5.0 for general sceanrios, although we do not want to noticably regress steady-state performance while addressing the stated goals including improving startup performance. It is possible some code-gen options described later may have steady-state performance gains, but again that is not a stated goal.
+
+<details>
+  <summary>5.0 master startup benchmarks (click to expand)</summary>
+
+A small POCO was used (`LoginViewModel` benchmark test class).
+
+| Serializer |  Serialize (us)  | Serialize ratio | Deserialize (us) | Deserialize ratio |
+| :-- | :-- | :-- | :-- | :--
+| **System.Text.Json** | 4,720 | 1.00 | 19,379 | 1.00 
+| **Json.NET** | 24,916 | 5.28 | 102,226 | 5.28 
+| **Utf8Json** | 7,290 | 1.54 | 106,817 | 5.51 
+| **Jil** | 66,456 | 14.08 | 170,548 | 8.80 
+</details>
+
+## Reduced memory usage
+Reducing private bytes (non-shared per-process) memory is important for Bing scenarios and other scenarios that have many process using the same POCOs or start and stop many processes. Shared memory can be amortized across many processes plus is already there when a new processes starts and thus requires no additional overhead to generate.
+
+This can be done by:
+- Avoiding runtime code generation including Reflection.Emit or JITting.
+- Capturing POCO metadata in generated code instead of RAM.
+
+## Reduced size-on-disk
```

Comment:
Can we quantify this?

Reply:
I assume you mean minimizing RAM\heap usage? Or do you mean size-on-disk? I do not have numbers yet for either but can get some numbers depending on what you find useful.

I do know that with large numbers of POCOs + many properties + longer-than-you-think property names (which is common) the cached metadata becomes significant, although the memory used is probably less than what the CLR maintains for each Type. Also FWIW currently for every property the serializer tracks 3 variants of its name - which I am currently looking at removing 1.